Add submit checkpoint functionality

jakovmitrovski opened the pull request:
- Add relayer that periodically submits a checkpoint to bitcoin for every existing subnet
- Implement submit_checkpoint functionality in ipc_lib which creates an OP_RETURN where checkpoint hash is submitted

jakovmitrovski committed:
Add submit checkpoint functionality - missing subnetPK signing

- Add relayer that periodically submits a checkpoint to bitcoin for every existing subnet
- Implement submit_checkpoint functionality in ipc_lib which creates an OP_RETURN where checkpoint hash is submitted

jakovmitrovski committed:
Add submit checkpoint functionality - missing subnetPK signing

- Add relayer that periodically submits a checkpoint to bitcoin for every existing subnet
- Implement submit_checkpoint functionality in ipc_lib which creates an OP_RETURN where checkpoint hash is submitted

OrestisAlpos commented on src/bin/relayer.rs:
```diff
@@ -33,6 +33,28 @@
+
+fn checkpoint() {
+    loop {
+        let subnets = IPCState::load_all().unwrap_or_else(|_| Vec::new());
```
The relayer is a separate entity, it runs on a machine unrelated to the machine that runs the bitcoin full node and maintains the ipc state. For example,  it can be a third party that's incentivized (paid) to monitor the subnets and checkpoint their state. So, technically, the relayer does not have access to these json files were we keep the ipc state.  For this demo it's ok to have the relayer read these files, as long as it is clear to us it is a simplification.

But what's more important is that we cannot do `SubnetState::new().get_checkpoint()`.  The checkpoint should be obtained by calling the `get_checkpoint()` function of the public interface of the subnet_simulator. And, in order to make things simpler and less dependent, we can make the subnet url a parameter of the relayer binary, so that it knows which subnet (subnet simulator in this Stage) to connect to, without having to know the ipc_state for all subnets.

OrestisAlpos commented on src/bitcoin_utils.rs:
```diff
@@ -516,3 +513,60 @@ pub fn write_arbitrary_data(
+        .unwrap();
+    if !signed_raw_transaction.complete {
+        println!("{:#?}", signed_raw_transaction.errors);
+        panic!("Transaction couldn't be signed.")
```
Can we avoid panic? Let's make this function return an error to the caller, the caller can then decide what do in case of an error.

OrestisAlpos committed:
just some text changes

OrestisAlpos committed:
an alternative design for SubnetInteractor and SubnetSimulator. The SubnetInteractor does not need to know about the SubnetState struct. It only talks to a subnet (a subnet simulator in our case) through its exposed interface, the SubnetState is a structure internal to that subnet. It does not need a dependency on the IPCState structure as well. It is enough if it knows how to contact the subnet validators (in our case, the single subnet simulator).

OrestisAlpos committed:
delete contents of main

jakovmitrovski committed:
Improve the workflow

- Add loading the subnets to interact with them
- Add validator pk when creating validator data
- Modify the parse commands logic
- Streamline the whole demo into main.rs
- Update README.md

jakovmitrovski committed:
Update relayer, checkpointing functionality and merge with parent branch

jakovmitrovski committed:
merge

jakovmitrovski committed:
Merge branch 'feature/interact' into feature/checkpoint

jakovmitrovski committed:
Fully implement checkpoint functionality done by relayer and signing SegWit transactions with SubnetPK

- Implement signing Segwit transactions with SubnetPK
- Refactor create command to generate and store a keypair representing the subnetPK
- Store the subnet_address rather than the public key in the ipc state
- Modify the workflow to account for the mentioned changes
- Update btc monitor to listen to checkpoints

jakovmitrovski committed:
Modify main.rs to utilize the .sh scripts. Replace SegwitV0 transactions with Taproot.

- Add scripts folder containing the .sh files to run the demo on ubuntu and macos and also to run the subnet interactor for a specific file
- Replace the subnetPK address to be a taproot address which allows the subnetPK to spend UTXOs using taproot and shnorr sigs.
- Make necessary changes to btc monitor when parsing the checkpoint (parse the string first, then hex-encode the checkpoint to display it)
- Modify the hash function to return bytes
- Modify the checkpoint transaction to directly append bytes in the op_return instead of converting utf-8 hex to bytes (this allows for longer subnet names)

jakovmitrovski committed:
Merge branch 'main' into feature/checkpoint

jakovmitrovski committed:
resolve conflicts

jakovmitrovski committed:
error handling refactoring

OrestisAlpos committed:
small changes
